src/main/kotlin/eu/europa/ec/eudi/pidissuer/adapter/out/diploma/IssueJwtVcJsonDiploma.kt

```diff
@@ -0,0 +1,188 @@
+        credentialSubject = listOf(
+            AchievedClaim.attribute,
+            EntitledToClaim.attribute,
+            PerformedClaim.attribute,
+        ),
+    ),
+)
+
+class IssueJwtVcJsonDiploma(
+    val credentialIssuerId: CredentialIssuerId,
+    private val generateUserDiplomaData: GenerateUserDiplomaData,
+    private val notificationsEnabled: Boolean,
+    private val generateNotificationId: GenerateNotificationId,
+    private val extractJwkFromCredentialKey: ExtractJwkFromCredentialKey,
+    private val issuerSigningKey: IssuerSigningKey,
+    private val clock: Clock,
+    private val storeIssuedCredential: StoreIssuedCredential,
+) : IssueSpecificCredential<JsonElement> {
+
+    private val log = LoggerFactory.getLogger(IssueJwtVcJsonDiploma::class.java)
+
+    override val supportedCredential: JwtVcJsonCredentialConfiguration
+        get() = IssueJwtVcJsonDiplomaConfiguration
+
+    override val publicKey: JWK? = null
+
+    private val validateProof = ValidateProof(credentialIssuerId)
+
+    context(Raise<IssueCredentialError>)
+    override suspend fun invoke(
+        authorizationContext: AuthorizationContext,
+        request: CredentialRequest,
+        credentialIdentifier: CredentialIdentifier?,
+        expectedCNonce: CNonce,
+    ): CredentialResponse<JsonElement> = coroutineScope {
+        log.info("Handling issuance request ...")
+        val holderPubKey = async(Dispatchers.Default) { holderPubKey(request, expectedCNonce) }
+
+        val jwt = encodeDiplomaToJwtVcJson(holderPubKey.await())
+
+        val notificationId =
+            if (notificationsEnabled) generateNotificationId()
+            else null
+
+        storeIssuedCredential(
+            IssuedCredential(
+                format = JWT_VC_JSON_FORMAT,
+                type = supportedCredential.credentialDefinition.type.joinToString(" "),
+                holder = holderPubKey.await().toPublicJWK().computeThumbprint().toString(),
+                holderPublicKey = holderPubKey.await().toPublicJWK(),
+                issuedAt = clock.instant(),
+                notificationId = notificationId,
+            ),
+        )
+
+        CredentialResponse.Issued(JsonPrimitive(jwt), notificationId)
+            .also {
+                log.info("Successfully issued diploma")
+                log.debug("Diploma data {}", it)
+            }
+    }
+
+    private fun encodeDiplomaToJwtVcJson(holderKey: JWK): String {
+        val _now = clock.instant()
+
+        val (diplomaName, achieved, entitledTo, performed) = generateUserDiplomaData()
+
+        val credential = buildCredential {
+            credentialMetadata {
+                type(diplomaName)
+                issuer(credentialIssuerId.value.toString())
+                issueDate(_now)
+                expirationDate(_now.plusSeconds(86400))
+            }
+            credentialSubject {
+                val didUrl = createDidUrl(holderKey, DidMethod.KEY).getOrThrow()
```

Comment:
Whose didUrl is this? Better provide a business name.

Also given that DID and DID URL are being representing with the same class (`DIDURL`), personally I would prefer something more concrete.

Semantics seems not clear for the `holderPubKey` parameter.